### PR TITLE
[spec] DataCount section, syntax, and validation

### PR DESCRIPTION
**Pull request by binji**
Fixes issue #73.

### CODE REVIEW

**Comment by binji** [comment]
Rendered version here: https://webassembly.github.io/bulk-memory-operations/core/

File: document/core/binary/modules.rst [hunk 1]
```diff
@@ -479,9 +507,11 @@ The lengths of vectors produced by the (possibly empty) :ref:`function <binary-f
      \Bcustomsec^\ast \\ &&&
      \elem^\ast{:\,}\Belemsec \\ &&&
      \Bcustomsec^\ast \\ &&&
+     \X{dc}^?{:\,}\Bdatacountsec \\ &&&
+     \Bcustomsec^\ast \\ &&&
      \X{code}^n{:\,}\Bcodesec \\ &&&
      \Bcustomsec^\ast \\ &&&
-     \data^\ast{:\,}\Bdatasec \\ &&&
+     \data^{\X{dc'}}{:\,}\Bdatasec \\ &&&
```
Comment (rossberg) [inline document/core/binary/modules.rst]: Nit: use metavariable *m*, analogous to *n* for func/code section

File: document/core/syntax/modules.rst [hunk 2]
```diff
@@ -21,6 +22,7 @@ and provide initialization in the form of :ref:`active <syntax-active>` and :ref
      \MGLOBALS~\vec(\global), \\&&&&
      \MELEM~\vec(\elem), \\&&&&
      \MDATA~\vec(\data), \\&&&&
+     \MDATACOUNT~\u32^?, \\&&&&
```
Comment (rossberg) [inline document/core/syntax/modules.rst]: Hm, why do you need to extend the abstract syntax (and validation)? The side condition is already handled as a detail of the binary format, so can't the rest remain as before? IIRC, there was no corresponding change in the interpreter.

Comment (binji) [reply document/core/syntax/modules.rst]: We also need to fail when `memory.init` or `data.drop` are used, but the data count section is not present. I forgot that we handled this during decoding -- I'll update it tomorrow.

File: document/core/binary/modules.rst [hunk 3]
```diff
@@ -433,6 +434,32 @@ It decodes into a vector of :ref:`data segments <syntax-data>` that represent th
...
+It decodes into an optional :ref:`u32 <syntax-uint>` that represents the number of :ref:`data segments <syntax-data>` in the :ref:`data section <binary-datasec>`. If this count does not match the length of the data segment vector, the module is malformed.
+
+.. math::
+   \begin{array}{llclll}
+   \production{data count section} & \Bdatacountsec &::=&
+     \X{n}^?{:}\Bsection_{12}(\Bu32) &\Rightarrow& \X{n}^? \\
+   \end{array}
+
+.. note::
+  The data count section is used to simplify single-pass validation. Since the
```
[the remark this answers is not in the document]
Comment (binji) [reply document/core/binary/modules.rst]: Done.

File: document/core/binary/modules.rst (hunk 1 above)
Comment (binji) [reply document/core/binary/modules.rst]: Done.

File: document/core/syntax/modules.rst (hunk 2 above)
Comment (binji) [reply document/core/syntax/modules.rst]: Done.

**Comment by binji** [comment]
@rossberg Gonna land this. Not quite sure about the `\iff` condition on the module, but feel free to add a comment if I did it wrong and I'll address in another PR. :-)

**Pull request merged**